ui/src/routes/devices.$id.settings.macros.tsx

```diff
@@ -0,0 +1,1439 @@
+                      const newErrors = { ...errors };
+                      delete newErrors.description;
+                      setErrors(newErrors);
+                    }
+                  }}
+                  placeholder="Description (optional)"
+                />
+                <ErrorMessage error={errors.description} />
+              </div>
+            </div>
+            
+            <div className="mt-4">
+              <div className="macro-section-header">
+                <label className="macro-section-title">
+                  Steps:
+                </label>
+                <span className="macro-section-subtitle">
+                  {newMacro.steps?.length || 0}/{MAX_STEPS_PER_MACRO} steps
+                </span>
+              </div>
+              {errors.steps && errors.steps[0]?.keys && (
+                <div className="mt-2">
+                  <ErrorMessage error={errors.steps[0].keys} />
+                </div>
+              )}
+              <div className="mt-2 text-xs text-slate-500 dark:text-slate-400">
+                You can add up to {MAX_STEPS_PER_MACRO} steps per macro
+              </div>
+              <div className="mt-2 space-y-4">
+                {(newMacro.steps || []).map((step, stepIndex) => (
+                  <MacroStepCard
+                    key={stepIndex}
+                    step={step}
+                    stepIndex={stepIndex}
+                    onDelete={newMacro.steps && newMacro.steps.length > 1 ? () => {
+                      const newSteps = [...(newMacro.steps || [])];
+                      newSteps.splice(stepIndex, 1);
+                      setNewMacro(prev => ({ ...prev, steps: newSteps }));
+                    } : undefined}
+                    onMoveUp={() => {
+                      const newSteps = handleStepMove(stepIndex, 'up', newMacro.steps || []);
+                      setNewMacro(prev => ({ ...prev, steps: newSteps }));
+                    }}
+                    onMoveDown={() => {
+                      const newSteps = handleStepMove(stepIndex, 'down', newMacro.steps || []);
+                      setNewMacro(prev => ({ ...prev, steps: newSteps }));
+                    }}
+                    isDesktop={isDesktop}
+                    onKeySelect={(option) => handleKeySelect(stepIndex, option)}
+                    onKeyQueryChange={(query) => handleKeyQueryChange(stepIndex, query)}
+                    keyQuery={keyQueries[stepIndex] || ''}
+                    getFilteredKeys={() => getFilteredKeys(stepIndex)}
+                    onModifierChange={(modifiers) => handleModifierChange(stepIndex, modifiers)}
+                    onDelayChange={(delay) => handleDelayChange(stepIndex, delay)}
+                    isLastStep={stepIndex === (newMacro.steps?.length || 0) - 1}
+                  />
+                ))}
+
+                <div className="mt-4 border-t border-slate-200 pt-4 dark:border-slate-700">
+                  <button
```

Comment:
Use existing Button Component

Reply:
converted everything over that i could see, lmk if i missed a shared component